File: Workout_tracker/vpulz_mobile/src/shared/ai/AIService.ts
type AIContext = {
  pendingSync: number;
  activeWorkoutSummary: string;
  recentSummary: string;
};

type ProviderResult = {
  provider: 'groq' | 'gemini' | 'local';
  answer: string;
};

const REQUEST_TIMEOUT_MS = 12000;

function buildPrompt(question: string, context: AIContext): string {
  return [
    'You are VPULZ AI coach.',
    'Give short, practical training advice in 2-5 bullets.',
    `Pending sync actions: ${context.pendingSync}`,
    `Active workout: ${context.activeWorkoutSummary}`,
    `Recent history: ${context.recentSummary}`,
    `Question: ${question}`,
  ].join('\n');
}

async function fetchWithTimeout(url: string, init: RequestInit, timeoutMs = REQUEST_TIMEOUT_MS): Promise<Response> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);

  try {
    return await fetch(url, {
      ...init,
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timeout);
  }
}

function isRetryableStatus(status: number): boolean {
  return status === 408 || status === 409 || status === 425 || status === 429 || status >= 500;
}

function isAbortError(error: unknown): boolean {
  return error instanceof Error && error.name === 'AbortError';
}

async function parseResponseError(response: Response, providerName: string): Promise<Error> {
  const text = await response.text();
  const detail = text ? text.slice(0, 180) : `${providerName} request failed (${response.status})`;
  return new Error(detail);
}

async function withTransientRetry(fn: () => Promise<string>): Promise<string> {
  try {
    return await fn();
  } catch (error) {
    if (isAbortError(error)) {
      return fn();
    }

    const message = error instanceof Error ? error.message : '';
    const maybeRetryable = /429|5\d\d|timeout|timed out|temporarily|rate/i.test(message);
    if (!maybeRetryable) {
      throw error;
    }

    return fn();
  }
}

async function groqProvider(question: string, context: AIContext): Promise<string> {
  const apiKey = process.env.EXPO_PUBLIC_GROQ_API_KEY;
  if (!apiKey) {
    throw new Error('Missing Groq API key');
  }

  const model = process.env.EXPO_PUBLIC_GROQ_MODEL || 'llama-3.1-8b-instant';
  const content = buildPrompt(question, context);

  const response = await fetchWithTimeout('https://api.groq.com/openai/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      messages: [
        { role: 'system', content: 'You are a performance-focused training coach.' },
        { role: 'user', content },
      ],
      temperature: 0.2,
      max_tokens: 300,
    }),
  });

  if (!response.ok) {
    if (isRetryableStatus(response.status)) {
      throw new Error(`Groq retryable ${response.status}`);
    }
    throw await parseResponseError(response, 'Groq');
  }

  const payload = (await response.json()) as {
    choices?: Array<{ message?: { content?: string } }>;
  };

  const answer = payload.choices?.[0]?.message?.content?.trim();
  if (!answer) {
    throw new Error('Groq empty response');
  }

  return answer;
}

async function geminiProvider(question: string, context: AIContext): Promise<string> {
  const apiKey = process.env.EXPO_PUBLIC_GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error('Missing Gemini API key');
  }

  const model = process.env.EXPO_PUBLIC_GEMINI_MODEL || 'gemini-1.5-flash-8b';
  const response = await fetchWithTimeout(
    `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [
          {
            role: 'user',
            parts: [
              {
                text: buildPrompt(question, context),
              },
            ],
          },
        ],
        generationConfig: {
          temperature: 0.2,
          maxOutputTokens: 320,
        },
      }),
    }
  );

  if (!response.ok) {
    if (isRetryableStatus(response.status)) {
      throw new Error(`Gemini retryable ${response.status}`);
    }
    throw await parseResponseError(response, 'Gemini');
  }

  const payload = (await response.json()) as {
    candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>;
  };

  const answer = payload.candidates?.[0]?.content?.parts?.[0]?.text?.trim();
  if (!answer) {
    throw new Error('Gemini empty response');
  }

  return answer;
}

function localFallback(question: string, context: AIContext): string {
  const normalized = question.toLowerCase();

  if (normalized.includes('deload') || normalized.includes('fatigue')) {
    return 'Deload by 10-15% for one session and keep movement quality high before ramping load again.';
  }

  if (normalized.includes('increase') || normalized.includes('progress')) {
    return 'Increase one variable only (load +2.5kg or +1 rep) if your previous sets were technically clean.';
  }

  if (context.activeWorkoutSummary !== 'No active workout') {
    return `Finish your current session first, then adjust next block. ${context.activeWorkoutSummary}.`;
  }

  return 'Keep consistency first: hit planned volume this week, then progress next week.';
}

function getProviderOrder(): Array<'groq' | 'gemini'> {
  const raw = process.env.EXPO_PUBLIC_AI_PROVIDER_ORDER?.trim().toLowerCase();
  if (!raw) {
    return ['groq', 'gemini'];
  }

  const values = raw
    .split(',')
    .map((value) => value.trim())
    .filter((value): value is 'groq' | 'gemini' => value === 'groq' || value === 'gemini');

  if (!values.length) {
    return ['groq', 'gemini'];
  }

  return Array.from(new Set(values));
}

export async function getCoachResponse(question: string, context: AIContext): Promise<ProviderResult> {
  const providers = getProviderOrder();

  for (const provider of providers) {
    try {
      if (provider === 'groq') {
        const answer = await withTransientRetry(() => groqProvider(question, context));
        return { provider: 'groq', answer };
      }

      const answer = await withTransientRetry(() => geminiProvider(question, context));
      return { provider: 'gemini', answer };
    } catch {
      continue;
    }
  }

  return { provider: 'local', answer: localFallback(question, context) };
}

File: vpulz_platform/backend/services/ai_service.py
from __future__ import annotations

import json
from typing import Any
from urllib import error, request

from vpulz_platform.backend.utils.config import settings


def _truncate(text: str, limit: int = 2000) -> str:
    return text[:limit]


class AIService:
    def _prompt(self, question: str, context: dict[str, Any]) -> str:
        return _truncate(
            "\n".join(
                [
                    "You are VPULZ AI coach.",
                    "Give practical coaching guidance in short bullet points.",
                    f"Question: {question}",
                    f"Context: {json.dumps(context, ensure_ascii=False)}",
                ]
            )
        )

    def groq_provider(self, question: str, context: dict[str, Any]) -> str:
        if not settings.groq_api_key:
            raise RuntimeError("Missing GROQ_API_KEY")

        model = settings.groq_model
        body = {
            "model": model,
            "messages": [
                {"role": "system", "content": "You are a performance-focused strength coach."},
                {"role": "user", "content": self._prompt(question, context)},
            ],
            "temperature": 0.2,
            "max_tokens": 280,
        }

        req = request.Request(
            "https://api.groq.com/openai/v1/chat/completions",
            method="POST",
            data=json.dumps(body).encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {settings.groq_api_key}",
            },
        )

        with request.urlopen(req, timeout=8) as response:
            payload = json.loads(response.read().decode("utf-8"))

        answer = payload.get("choices", [{}])[0].get("message", {}).get("content", "").strip()
        if not answer:
            raise RuntimeError("Groq empty response")
        return answer

    def gemini_provider(self, question: str, context: dict[str, Any]) -> str:
        if not settings.gemini_api_key:
            raise RuntimeError("Missing GEMINI_API_KEY")

        model = settings.gemini_model
        url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={settings.gemini_api_key}"
        body = {
            "contents": [{"role": "user", "parts": [{"text": self._prompt(question, context)}]}],
            "generationConfig": {
                "temperature": 0.2,
                "maxOutputTokens": 300,
            },
        }

        req = request.Request(
            url,
            method="POST",
            data=json.dumps(body).encode("utf-8"),
            headers={"Content-Type": "application/json"},
        )

        with request.urlopen(req, timeout=8) as response:
            payload = json.loads(response.read().decode("utf-8"))

        answer = (
            payload.get("candidates", [{}])[0]
            .get("content", {})
            .get("parts", [{}])[0]
            .get("text", "")
            .strip()
        )
        if not answer:
            raise RuntimeError("Gemini empty response")
        return answer

    def fallback(self, question: str, context: dict[str, Any]) -> str:
        normalized = question.lower()
        if "deload" in normalized or "fatigue" in normalized:
            return "Take a 1-week deload: reduce load by 10-15% and keep movement quality high."
        if "progress" in normalized or "increase" in normalized:
            return "Progress one variable at a time: +1 rep or +2.5kg when reps stay clean."
        pending = context.get("pending_sync", 0)
        return f"Focus on execution quality today. Pending sync actions: {pending}."

    def answer(self, question: str, context: dict[str, Any]) -> dict[str, str]:
        providers = [provider.strip().lower() for provider in settings.ai_provider_order.split(",") if provider.strip()]
        if not providers:
            providers = ["groq", "gemini"]

        for provider in providers:
            try:
                if provider == "groq":
                    return {"provider": "groq", "answer": self.groq_provider(question, context)}
                if provider == "gemini":
                    return {"provider": "gemini", "answer": self.gemini_provider(question, context)}
            except (error.HTTPError, error.URLError, TimeoutError, RuntimeError, json.JSONDecodeError):
                continue

        return {"provider": "fallback", "answer": self.fallback(question, context)}
